TRD/TRDbase/AliTRDtrackerV1.cxx

```diff
@@ -4381,3 +4382,311 @@ Double_t AliTRDtrackerV1::AliTRDtrackFitterRieman::CalculateReferenceX(){
   }
   return fTracklets[startIndex]->GetX0() + (2.5 - startIndex) * meanDistance - 0.5 * (AliTRDgeometry::AmThick() + AliTRDgeometry::DrThick());
 }
+
+///
+/// \param esdTrack  - esdTrack to foolow
+/// \return
+/// Algorithm:
+/// loop over all possible TOF hits
+///     propagate Out parameters to the TOF hit
+///     apply chi2 cut
+///     update track with TOF hit
+///     create "TRDtrack" from updated track
+///
+///     loop overs TRD  layers
+///          create tracklet
+///          tracklet.AttachClusters
+///                   * finding clusters in road
+///          Refit tracklet - Robust
+///                   * QA information - extracting position and angle - to caompare with track angle
+///          calculated "causality information  to store per TOF hit
+///             bitmask,   Ncl array per layer
+///                   * chamber non active (-2) , dead zone (-1) , no hit (0),     Ncl compressed format  (e.g int(Ncl/4))
+///             material budget mask
+///                   * X/X0 per layer
+///    refit TRD track with tracklets
+///    Update ESD track if not TRD track provide by standard means  - to check in debug streamer
+Int_t           AliTRDtrackerV1::FollowInterpolationsTPCTOF(AliESDtrack &esdTrack){
+  //if (esdTrack.GetNumberOfTRDslices()<5) return 0; //TODO 0 debug line to remove
+  TTreeSRedirector *pstreamer = fkReconstructor->GetDebugStream(AliTRDrecoParam::kTracker);
+  enum {kHole=0x100,kBoundary=0x200};
+  const Float_t kStepSize=3;
+  const Float_t chi2Cut=49;
+  const Float_t vdriftCut=0.6;
+  const Float_t gainCut=0.8;
+  const double kBoundaryEps = 4;   // 4 cm dead zone
+  double boundaryEps = kBoundaryEps + AliTRDReconstructor::GetExtraBoundaryTolerance();
+  AliTRDcalibDB* const calibration = AliTRDcalibDB::Instance();
+  AliTRDtrackingChamber *chamber = NULL;
+  Double_t driftLength = .5*AliTRDgeometry::AmThick() + AliTRDgeometry::DrThick();
+  const AliTRDCalDet *exbDet = calibration->GetExBDet();
+  const AliTRDCalDet *t0Det=calibration->GetT0Det();
+  const AliTRDCalDet *vdDet=calibration->GetVdriftDet();
+  const AliTRDCalDet *gainDet=calibration->GetGainFactorDet();
+  //
+  Float_t gainMeanRobust=gainDet->GetMeanRobust(0.75);
+  Float_t gainRMSRobust=gainDet->GetRMSRobust(0.75);
+  Float_t gainRMS=gainDet->GetRMS();
+  Float_t exbMeanRobust=exbDet->GetMeanRobust(0.75);
+  Float_t exbRMSRobust=exbDet->GetRMSRobust(0.75);
+  Float_t exbRMS=exbDet->GetRMS();
+  Float_t t0MeanRobust=t0Det->GetMeanRobust(0.75);
+  Float_t t0RMSRobust=t0Det->GetRMSRobust(0.75);
+  Float_t t0RMS=t0Det->GetRMSRobust(0.75);
+  Float_t vdMeanRobust=vdDet->GetMeanRobust(0.75);
+  Float_t vdRMSRobust=vdDet->GetRMSRobust(0.75);
+  Float_t vdRMS=vdDet->GetRMSRobust(0.75);
```

Comment:
why do you need to do this per track rather than calculating all these robust estimators to only once? They do quite some calculations.

Reply:
it is not needed to do it per track. Could be cached.  Only HV is changing in time.
Time should be not big as it is only sort of 540 values
 
I can do caching  using static variables, keeping variable local in function.

Reply:
> Hello @miranov25
> 
> Thanks! Please see one comment inline about robust estimators. I cannot judge how efficient they will be to detect the status of the chamber, but there is not reason to do this per track.
> Otherwise, the changes look fine (cannot judge about the stability of the reconstruction with FollowInterpolationsTPCTOF, I hope you have tested it on reasonable statistics.

Concerning tests I was running 5 events x 120 chunks. No  failures.
 
> I see you've added here also the ITS PID using interpolated momentum, but I thought you agreed with @fprino to have a switch which is OFF by default, I don't see it.

I did not implement yet ITS switch yet, this I will do later today.

Reply:
sorting 540 values for every track is also non-negligible, moreover that you use only 2 out of many variables you compute.
I think it is worth at least to initialize them as static variables.